### ISSUE TITLE
Cannot authenticate using SASL PLAIN

### ISSUE THREAD

**Issue by stevo** [zendesk/ruby-kafka#448]
* Version of Ruby: 2.4.2
* Version of Kafka: 0.10
* Version of ruby-kafka: 0.5.0.beta3, 0.4.3

###### Steps to reproduce

```ruby
Kafka.new(seed_brokers: ["xx.xx.xx.xx:9092"], sasl_plain_username: 'kafka_user', sasl_plain_password: 'xxxxxxxxxxxxx').deliver_message("Hello, World!", topic: "test")
```

###### Expected outcome

```
Successfully appended 1 messages to test/0 on ec2-xx-xx-xx-xx.us-east-2.compute.amazonaws.com:9092 (node_id=132517971)
```

###### Actual outcome

```
New topics added to target list: test
Fetching cluster metadata from kafka://xx.xx.xx.xx:9092
Opening connection to xx.xx.xx.xx:9092 with client id ruby-kafka...
Sending request 1 to xx.xx.xx.xx:9092
Waiting for response 1 from xx.xx.xx.xx:9092
Received response 1 from xx.xx.xx.xx:9092
Failed to fetch metadata from kafka://xx.xx.xx.xx:9092: SASL PLAIN authentication failed: Operation timed out
Closing socket to xx.xx.xx.xx:9092
Kafka::ConnectionError: Could not connect to any of the seed brokers:
- kafka://xx.xx.xx.xx:9092: SASL PLAIN authentication failed: Operation timed out
```

Somehow during authentication (https://github.com/zendesk/ruby-kafka/blob/master/lib/kafka/sasl_plain_authenticator.rb#L29) decoder tried to read (https://github.com/zendesk/ruby-kafka/blob/master/lib/kafka/protocol/decoder.rb#L95) 0 (https://github.com/zendesk/ruby-kafka/blob/master/lib/kafka/protocol/decoder.rb#L41) bytes

When I've added following line before https://github.com/zendesk/ruby-kafka/blob/master/lib/kafka/protocol/decoder.rb#L95 -> `return '' if number_of_bytes == 0` I've managed to authenticate and deliver message successfully. Still I am not entirely sure if this fix is correct. Please help :)

**Comment by dasch** [comment]
Returning an empty string when asked to read zero bytes makes sense, but I'd like to understand why this happens – surely the SASL PLAIN protocol has the answer for what to expect.

**Comment by stevo** [comment]
Exactly @dasch ! Do you know where can we find some detailed specification of how communication between client and kafka looks like during SASL PLAIN authentication?

**Comment by dasch** [comment]
Not really, I didn't add that feature myself.

**Comment by stevo** [comment]
Maybe @trthomps can help here ? :)

**Comment by trthomps** [comment]
It's possible something changed in newer versions of Ruby because when I wrote this I was testing against a SASL PLAIN 0.11.x cluster using ruby 2.3.3.  I used the python library for inspiration:

https://github.com/dpkp/kafka-python/blob/34dc9dd2fe6b47f4542c5a131e0e0cbc1b00ed80/kafka/conn.py#L519

**Comment by stevo** [comment]
It seems that this `0` we get is something that should be expected when authentication succeedes...

https://github.com/dpkp/kafka-python/blob/34dc9dd2fe6b47f4542c5a131e0e0cbc1b00ed80/kafka/conn.py#L542

It seems it is also ok to read 0 bytes from socket. The timeout occurs rather during

```ruby
  unless IO.select([@socket], nil, nil, @timeout)
        raise Errno::ETIMEDOUT
      end
```

I understand that we're waiting for data here, but in fact we do not expect any data to be received. So I think that adding `return '' if num_bytes == 0` in `Kafka::SocketWithTimeout#read` will be better idea than doing the same in decoder. What do you think, @trthomps, @dasch ?

**Comment by dasch** [comment]
@stevo the decoder calls the socket, so if we make the decoder just return an empty string from `#bytes` when `size` is zero we should be good to go.

**Comment by dasch** [comment]
Even better would be implementing it in `Decoder#read`.

**Comment by stevo** [comment]
https://github.com/zendesk/ruby-kafka/pull/471

**Comment by dasch** [comment]
Fixed by #471.
